builtin/logical/totp/path_keys.go

```diff
@@ -0,0 +1,400 @@
+			return logical.ErrorResponse("an error occured while parsing url string"), nil
+		}
+
+		//Set up query object
+		urlQuery := urlObject.Query()
+		path := strings.TrimPrefix(urlObject.Path, "/")
+		index := strings.Index(path, ":")
+
+		//Read issuer
+		urlIssuer := urlQuery.Get("issuer")
+		if urlIssuer != "" {
+			issuer = urlIssuer
+		} else {
+			if index != -1 {
+				issuer = path[:index]
+			}
+		}
+
+		//Read account name
+		if index == -1 {
+			accountName = path
+		} else {
+			accountName = path[index+1:]
+		}
+
+		//Read key string
+		keyString = urlQuery.Get("secret")
+
+		//Read period
+		periodQuery, err := strconv.Atoi(urlQuery.Get("period"))
+		if err == nil {
+			period = periodQuery
+		}
+
+		//Read digits
+		digitsQuery, err := strconv.Atoi(urlQuery.Get("digits"))
+		if err == nil {
+			digits = digitsQuery
+		}
+
+		//Read algorithm
+		algorithmQuery := urlQuery.Get("algorithm")
+		if algorithmQuery != "" {
+			algorithm = algorithmQuery
+		}
+	}
+
+	// Translate digits and algorithm to a format the totp library understands
+	var keyDigits otplib.Digits
+	switch digits {
+	case 6:
+		keyDigits = otplib.DigitsSix
+	case 8:
+		keyDigits = otplib.DigitsEight
+	default:
+		return logical.ErrorResponse("the digit value can only be 6 or 8"), nil
+	}
+
+	var keyAlgorithm otplib.Algorithm
+	switch algorithm {
+	case "SHA1":
+		keyAlgorithm = otplib.AlgorithmSHA1
+	case "SHA256":
+		keyAlgorithm = otplib.AlgorithmSHA256
+	case "SHA512":
+		keyAlgorithm = otplib.AlgorithmSHA512
+	default:
+		return logical.ErrorResponse("the algorithm value is not valid"), nil
+	}
+
+	// Enforce input value requirements
+	if period <= 0 {
+		return logical.ErrorResponse("the period value must be greater than zero"), nil
+	}
+
+	switch skew {
+	case 0:
+	case 1:
+	default:
+		return logical.ErrorResponse("the skew value must be 0 or 1"), nil
+	}
+
+	if qrSize <= 0 {
+		return logical.ErrorResponse("the qr_size value must be greater than zero"), nil
+	}
+
+	if keySize <= 0 {
+		return logical.ErrorResponse("the key_size value must be greater than zero"), nil
+	}
+
+	// Period, Skew and Key Size need to be unsigned ints
+	uintPeriod := uint(period)
+	uintSkew := uint(skew)
+	uintKeySize := uint(keySize)
+
+	var response logical.Response
+
+	switch generate {
+	case true:
+		// If the key is generated, Account Name and Issuer are required.
+		if accountName == "" {
+			return logical.ErrorResponse("the account_name value is required for generated keys"), nil
+		}
+
+		if issuer == "" {
+			return logical.ErrorResponse("the issuer value is required for generated keys"), nil
+		}
+
+		// Generate a new key
+		keyObject, err := totplib.Generate(totplib.GenerateOpts{
+			Issuer:      issuer,
+			AccountName: accountName,
+			Period:      uintPeriod,
+			Digits:      keyDigits,
+			Algorithm:   keyAlgorithm,
+			SecretSize:  uintKeySize,
+		})
+		if err != nil {
+			return logical.ErrorResponse("an error occured while generating a key"), nil
```

Comment:
Same here, log the error or return it (in this case since the error wouldn't have the key value in it it's probably safe to just return) -- or both!